truffle/docs/Safepoints.md

```diff
@@ -0,0 +1,178 @@
+# Truffle Language Safepoint Tutorial
+
+Since 21.1 Truffle has support for guest language safepoints. 
+Truffle safepoints allow to interrupt the guest language execution to perform thread local actions submitted by a language or tool. 
+A safepoint is a location during the guest language execution where the state is consistent and other operations can read its state.
+
+This replaces previous instrumentation or assumption based approaches to safepoints, which required the code to be invalidated for a thread local action to be performed.
+The new implementation uses fast thread local access checks and the callee register saved stub calls to optimize for performance and keep the overhead minimal.
+This means that for every loop back-edge and method exit we perform an additional non-volatile read which can potentially lead to slight slow-downs.
+
+## Use Cases
+
+Common use-cases of Truffle language safepoints are:
+
+* Cancellation, requested exit or interruptions during guest language execution. The stack is unwound by submitting a thread local action.
+* Reading the current stack trace information for other thread than the currently executing thread.
+* Enumerating all object references active on the stack.
+* Implement guest languages that expose a safepoint mechanism as part of their development toolkit.
+* Debuggers evaluating expressions in languages that do not support execution on multiple threads.
+
+## Language Support
+
+Safepoints are explicitly polled by invoking the `TruffleSafepoint.poll(Node)` method. 
+A Truffle guest language implementation must ensure that a safepoint is polled repeatedly within a constant time interval. 
+For example, a single arithmetic expression completes within a constant number of CPU cycles. 
+However, a loop that summarizes values over an array uses a non-constant time dependent on the actual array size. 
+This typically means that safepoints are best polled at the end of loops and at the end of function or method calls to cover recursion. 
+In addition, any guest language code that blocks the execution, like guest language locks, need to use the  `TruffleSafepoint.setBlocked(Interrupter)` API to allow cooperative polling of safepoints while the thread is waiting.
+
+Please read more details what steps languages need to do in the [javadoc](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/TruffleSafepoint.html).
+
+## Thread Local Actions
```

Comment:
The difference between a 'safepoint' and a 'thread local action' isn't made clear in this document - could be confusing.